### ISSUE TITLE
 Formatting mode

### ISSUE THREAD

**Issue by LiamPattinson** [PlasmaFAIR/fortitude#3]
Alongside a linting mode,  we could also provide a formatting mode. This could work by walking through the `tree_sitter::Tree` and rewriting the code with fixed whitespace rules. This would hopefully be more robust than a purely regex-based formatter, as it would be easier to distinguish between formatting rules in different contexts:

```f90
! Assignment operator should have space either side
myvar = x + y
! ... but not for keyword arguments
character(len=20, kind=c_char) :: string

! Colon should have a space after it
use mymodule, only: myfunc
! ... except when slicing
array(1:6) = 1
```

The following snippet can walk through the code via `tree_sitter` and print each token with spaces between. Some nodes appear to print twice, particularly those composed of multiple words (`double precision` goes to `doubleprecision doubleprecision`, `end module` goes to `endmodule endmodule`), but otherwise it reproduces simple codes well enough:

```rust
fn walk_and_print(node: &tree_sitter::Node, src: &str) {
    let mut cursor = node.walk();
    let mut children = node.children(&mut cursor);
    if children.len() > 0 {
        for child in children {
            walk_and_print(&child, &src);
        }
    } else {
        if node.is_named() {
            print!("{} ", node.utf8_text(src.as_bytes()).unwrap());
        } else {
            print!("{} ", node.kind());
        }
    }
}
```

To turn this into a formatter, it should be sufficient to implement a collection of functions for each node kind, and to call these via `match` instead of the simple recursion above. These could track indentation level as they traverse the tree, and different contexts could implement different rules for handling 'anonymous nodes' (mostly punctuation tokens).

Possible complications that are immediately apparent (I'm sure there will be more that I can't even imagine now):

- Maximum line length wrapping could be tricky, and there are multiple ways it could be achieved.
- Fortran has a lot of strange legacy options. It'll be no easy task to track down and understand all of them.
- [Edit] Will need to preserve newlines that the code author inserted to improve readability, but limit this to one newline.

**Comment by ZedThree** [comment]
I've got a proof-of-concept/WIP topiary config: https://github.com/PlasmaFAIR/topiary/tree/fortran

Some issues that would need addressing:

- tree-sitter doesn't see `end if` etc as separate tokens/nodes, so topiary is unable to whitespace them
- topiary won't handle lower/uppercasing keywords
- I don't think topiary handles long-lines (although it can do multi-line stuff), so we'd need to handle that ourselves

**Comment by LiamPattinson** [comment]
Nice! I don't think it's a huge problem if Topiary can't do everything, as a lot of the remaining issues could be fixed with easy find-and-replace regex such as (sed-style) `s/end\s*if/end if/g`, although we would need to be careful not to touch anything inside strings or comments, and it would be difficult to handle something awful like:

```f90
if (x) then
  ...
end &
  if
```

(Although given that this confuses Github's syntax highlighting, I'm not sure how likely we are to find code like this in practice)

**Comment by LiamPattinson** [comment]
Does topiary have support for `! NOFORMAT` guards (or however we would write them), or is that something we would have to implement ourselves?

**Comment by ZedThree** [comment]
Some interesting links on formatting, including how to handle line breaks: 
https://news.ycombinator.com/item?id=22706242

We almost certainly want to remove line continuations before doing any formatting. Although they don't appear in the AST, we can just search for `&` and then use the tree-sitter function to get node at given offset to check it's not in a comment or string literal, before merging lines.

**Comment by ZedThree** [comment]
User settings is a big one -- what's our philosophy going to be? Personally, I'm very much in favour of having little-to-no configuration, like `black`. `lfortran fmt` has number of spaces to indent (default 4), and whether to indent contents of procedures, programs, and modules.

**Comment by LiamPattinson** [comment]
There are only a few user settings I would like to see:

- Indent (I'd lean towards default 2, but it's not a strong conviction)
- Line length
- Lower/uppercase keywords (I'd say leave them as they are by default, but I know some style guides suggest one or the other)

fprettify has a lot of options for controlling whitespace:

```
 -w, --whitespace {0,1,2,3,4}
                        Presets for the amount of whitespace - 0: minimal
                        whitespace | 1: operators (except arithmetic),
                        print/read | 2: operators, print/read, plus/minus | 3:
                        operators, print/read, plus/minus, muliply/divide | 4:
                        operators, print/read, plus/minus, muliply/divide,
                        type component selector (default: 2)
  --whitespace-comma [WHITESPACE_COMMA]
                        boolean, en-/disable whitespace for comma/semicolons
                        (default: None)
  --whitespace-assignment [WHITESPACE_ASSIGNMENT]
                        boolean, en-/disable whitespace for assignments
                        (default: None)
  --whitespace-decl [WHITESPACE_DECL]
                        boolean, en-/disable whitespace for declarations
                        (requires '--enable-decl') (default: None)

... and bunch of similar rules follow
```

It can also modify cases with:

```
  --case CASE CASE CASE CASE
                        Enable letter case formatting of intrinsics by
                        specifying which of keywords, procedures/modules,
                        operators and constants (in this order) should be
                        lowercased or uppercased - 0: do nothing | 1:
                        lowercase | 2: uppercase (default: [0, 0, 0, 0])
```


I don't think we should provide this much flexibility, but I could potentially be swayed on that.

fprettify also has options to replace old-style relational operators, which we could do by borrowing functionality from our style rule, but one could argue that `check --select=style --fix` obviates the need for `format` to handle this. That would also be a way to handle rules like `keywords-missing-space` which already has some fine-tuning options under `[check.keyword-whitespace]` -- if `format` mode were to also apply these fixes, I'm not sure where we would put the options in the config file.

> `lfortran fmt` has ... whether to indent contents of procedures, programs, and modules.

I think our equivalent to that option should be, "yes, please indent your code".

**Comment by ZedThree** [comment]
@LiamPattinson Yeah, I think certainly much much fewer options than fprettify. And agreed that we should always indent.

Also agreed that we can just reuse a bunch of the style rules in formatting mode. I quite like the idea of `format` being opinionated and just unconditionally applying all the format-related style rules. Then if a project doesn't like `format` but does want to apply a subset of the style rules, they can just use `check --fix` with those.

@imciner2 My preferred style is something like "indent relative to the previous nesting", so a big complicated expression would look like:

```f90
variable = foo &
  + bar &
  * (thing &
    + (jaff &
       + zang) &
    - yeet) &
  + huff
```

(with really short line length for illustration purposes only!)

This is similar to what `black` does for Python. Is this what you meant by your second option?
